rotkehlchen/externalapis/alchemy.py

```diff
@@ -0,0 +1,400 @@
+        """
+        try:
+            from_asset = from_asset.resolve_to_asset_with_oracles()
+            to_asset = to_asset.resolve_to_asset_with_oracles()
+        except UnknownAsset as e:
+            raise PriceQueryUnsupportedAsset(e.identifier) from e
+
+        if (price_cache_entry := GlobalDBHandler.get_historical_price(
+            from_asset=from_asset,
+            to_asset=to_asset,
+            timestamp=timestamp,
+            max_seconds_distance=HOUR_IN_SECONDS * 6,
+            source=HistoricalPriceOracle.ALCHEMY,
+        )):
+            return price_cache_entry.price
+
+        try:
+            asset_identifier = self._get_asset_identifier(from_asset)
+        except UnsupportedAsset as e:
+            log.warning(
+                f'Tried to query alchemy historical price from {from_asset.identifier} '
+                f'to {to_asset.identifier}. But from_asset is not supported in alchemy',
+            )
+            raise NoPriceForGivenTimestamp(
+                from_asset=from_asset,
+                to_asset=to_asset,
+                time=timestamp,
+            ) from e
+
+        payload: dict[str, str | int] = {
+            'startTime': timestamp - HOUR_IN_SECONDS,
+            'endTime': timestamp + HOUR_IN_SECONDS,
+            'interval': '1h',
+        }
+        result = self._query(
+            endpoint='historical',
+            payload={**payload, **asset_identifier},
+        )
+        if (usd_price := self._deserialize_historical_price(
+            result=result,
+            asset_identifier=asset_identifier,
+            from_asset=from_asset,
+            timestamp=timestamp,
+            to_asset=to_asset,
+        )) == ZERO_PRICE:
+            raise NoPriceForGivenTimestamp(
+                from_asset=from_asset,
+                to_asset=to_asset,
+                time=timestamp,
+                rate_limited=False,
+            )
+
+        if to_asset != A_USD:
+            # we need to query intermediate price in this case.
+            to_asset_price = PriceHistorian().query_historical_price(
+                from_asset=A_USD,
+                to_asset=to_asset,
+                timestamp=timestamp,
+            )
+            price = Price(usd_price * to_asset_price)
+        else:
+            price = usd_price
+
+        GlobalDBHandler.add_historical_prices(entries=[HistoricalPrice(
+            from_asset=from_asset,
+            to_asset=to_asset,
+            source=HistoricalPriceOracle.ALCHEMY,
+            timestamp=timestamp,
+            price=price,
+        )])
+        return price
+
+    def can_query_history(
+            self,
+            from_asset: Asset,  # pylint: disable=unused-argument
+            to_asset: Asset,
+            timestamp: Timestamp,
+            seconds: int | None = None,
+    ) -> bool:
+        return not self.is_penalized()
+
+    @staticmethod
+    def _get_asset_identifier(asset: AssetWithOracles) -> dict[str, str]:
```

Comment:
Please rename this method. In our code `asset_identifier` is something else. Also why return a dict? This is a named tuple that this function returns. Network / address. As such I would name this something like `_alchemy_asset_data`

Moreover with a named_tuple you can have a function that implements this logic on the named tuple itself.

```python
        if 'symbol' in asset_identifier:
            result = self._query(
                payload={'symbols': asset_identifier['symbol']},
                endpoint='by-symbol',
            )
        else:
            result = self._query(
                payload={'addresses': [asset_identifier]},
                endpoint='by-address',
            )
```